### PR TITLE
Wire up trace uniter hooks

### PR DESCRIPTION
**Pull request by SimonRichardson**
Adds the tracer to the env hooks. Passes over the trace and span ids.
Trace flags will have to be done later once #16575 has landed.

This facilitates the ability to trace from juju unit agents through
the charm code. By just adding the `JUJU_TRACE_ID` and `JUJU_SPAN_ID`
any language or framework can pick those up and work on a trace
using that as the parent.

Most of the code is just pushing `context.Context` everywhere.

## Sceenshot

From this, you can see it leaving the juju-unit and then being picked up
via the juju controller charm.

See: https://github.com/juju/juju-controller/pull/55

Screenshot from 2024-01-29 11-37-45

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing

### Setting up tempo

It assumes that docker (docker-compose) is correctly installed:

```sh
$ git clone https://github.com/grafana/tempo.git
$ cd tempo/example/docker-compose/local
$ docker compose up -d
```

### Setting up juju-controller charm

```sh
$ cd <location of juju-controller charm>
$ git checkout <https://github.com/juju/juju-controller/pull/55>
$ charmcraft pack
```

### Bootstrapping

```sh
$ juju bootstrap lxd test --build-agent --config="open-telemetry-enabled=true" --config="open-telemetry-insecure=true" --config="open-telemetry-endpoint=192.168.0.60:4317" --config="open-telemetry-sample-ratio=1.0" --controller-charm-path=<location of juju-controller charm>/juju-controller_ubuntu-22.04-amd64-aarch64-arm64-s390x_ubuntu-20.04-amd64-aarch64-arm64-s390x.charm
```

## Links

**Jira card:** JUJU-[XXXX]

### CODE REVIEW

File: internal/worker/uniter/runner/runner.go [hunk 1]
```diff
@@ -316,6 +319,22 @@ func (runner *runner) runJujuExecAction() (err error) {
...
+	go func() {
+		defer cancel()
+
+		select {
+		case <-ctx.Done():
+		case <-abort:
+		}
+	}()
+	return c
+}
```
Comment (SimonRichardson) [inline internal/worker/uniter/runner/runner.go:336]: The consequence of mixing action abort with context. I wonder if there is a better pattern to handle this without storing a context on the action itself.

File: internal/worker/uniter/runner/context/context.go [hunk 2]
```diff
@@ -1405,12 +1409,19 @@ func (ctx *HookContext) HookVars(
 		return nil, errors.Trace(err)
 	}
 
+	if scope := coretrace.SpanFromContext(ctx).Scope(); scope.TraceID() != "" {
+		vars = append(vars,
+			"JUJU_TRACE_ID="+scope.TraceID(),
+			"JUJU_SPAN_ID="+scope.SpanID(),
+		)
+	}
```
Comment (SimonRichardson) [inline internal/worker/uniter/runner/context/context.go:1417]: The actual change.

Note: we should pass in if the trace is being recorded, but I need to change some code around to support that.

**Review by manadart — APPROVED** [review]
Nice. Works great.

**Pull request merged**